src/olaf/simplpedpop.rs

```diff
@@ -0,0 +1,976 @@
+
+        secret_commitments.insert(own_secret_commitment.compress().0);
+
+        let mut index = 0;
+        for message in &secret_commitments {
+            if message == &own_secret_commitment.compress().0 {
+                break;
+            }
+            index += 1;
+        }
+
+        for i in 0..secret_commitments.len() {
+            let input = Sha512::new().chain(scalar.as_bytes()).chain(i.to_string());
+            let random_scalar = Scalar::from_hash(input);
+            others_identifiers.insert(Identifier(random_scalar));
+        }
+
+        let own_identifier = *others_identifiers
+            .iter()
+            .nth(index)
+            .expect("This never fails because the index < len");
+        others_identifiers.remove(&own_identifier);
+
+        for (id, message) in others_identifiers.iter().zip(round1_public_messages_set) {
+            round1_public_messages.insert(*id, message);
+        }
+
+        let identifiers = Identifiers::new(own_identifier, others_identifiers);
+
+        (identifiers, round1_public_messages)
+    }
+
+    fn verify_round1_messages(
+        round1_public_messages: &BTreeSet<round1::PublicMessage>,
+    ) -> DKGResult<Vec<PublicKey>> {
+        let len = round1_public_messages.len();
+        let mut public_keys = Vec::with_capacity(len);
+        let mut proofs_of_possession = Vec::with_capacity(len);
+
+        // The public keys are the secret commitments of the participants
+        for round1_public_message in round1_public_messages {
+            let public_key = PublicKey::from_point(
+                *round1_public_message
+                    .secret_polynomial_commitment
+                    .coefficients_commitments
+                    .first()
+                    .expect("This never fails because the minimum threshold of the protocol is 2"),
+            );
+            public_keys.push(public_key);
+            proofs_of_possession.push(round1_public_message.proof_of_possession);
+        }
+
+        verify_batch(
+            vec![Transcript::new(b"Proof of Possession"); len],
+            &proofs_of_possession[..],
+            &public_keys[..],
+            false,
+        )
+        .map_err(DKGError::InvalidProofOfPossession)?;
+
+        Ok(public_keys)
+    }
+
+    fn generate_messages(
+        round2_public_data: &PublicData,
+        secret_polynomial: &SecretPolynomial,
+        secret_key: SecretKey,
+        secret_commitment: &SecretCommitment,
+    ) -> DKGResult<Messages> {
+        let mut private_messages = BTreeMap::new();
+
+        let enc_keys: Vec<RistrettoPoint> = round2_public_data
+            .round1_public_messages
+            .values()
+            .map(|msg| {
+                *msg.secret_polynomial_commitment
+                    .coefficients_commitments
+                    .first()
+                    .expect("This never fails because the minimum threshold of the protocol is 2")
+            })
+            .collect();
```

Comment:
I see..  You added an extra round so everyone could encrypt to everyone else's `f(0) G`.

Any idea what blockstream did here?  We must treat some input as "PKI", either
1. we save ourselves a round by making the user supply encryption keys initially, or else 
2. make the user provide the round 1 messages from the right people.

I think 1 looks more robust, in the sense that if they provide us a bad public key, then they're clearly buffoons, while 2 requires they do some checking all by themselves.

Reply:
Blockstream follows 1. 

The idea is that you do not really save a round because you still need to somehow know the encryption keys of the participants, you just save a round in _this_ protocol. 

And because of that, it excludes use cases where there is no way of knowing those keys beforehand, or they do not exist. 

That being said, I do not have a concrete use case and maybe it is not relevant.